src/controllers/action-queue/types/pushApiTypes.js

```diff
@@ -0,0 +1,113 @@
+// @flow
+
+import { type Cleaner, asArray, asBoolean, asDate, asEither, asJSON, asNull, asObject, asOptional, asString, asUnknown, uncleaner } from 'cleaners'
+
+import { asBase64 } from '../../../util/cleaners/asBase64'
+import { asBroadcastTx, asNewPushEvent, asPushEventState, asPushMessage, asPushTrigger } from './pushCleaners'
+import { type NewPushEvent } from './pushTypes'
+
+// ---------------------------------------------------------------------------
+// Request types
+// ---------------------------------------------------------------------------
+
+export type PushRequestBody = {
+  // The request payload:
+  data?: mixed,
+
+  // Who is making the request:
+  apiKey: string,
+  deviceId: string,
+  deviceToken?: string,
+
+  // For logins:
+  loginId?: Uint8Array
+}
+
+export type DeviceUpdatePayload = {
+  loginIds: Uint8Array[],
+  createEvents?: NewPushEvent[],
+  removeEvents?: string[]
+}
+
+// ---------------------------------------------------------------------------
+// Request cleaners
+// ---------------------------------------------------------------------------
+
+export const asPushRequestBody: Cleaner<PushRequestBody> = asObject({
+  // The request payload:
+  data: asUnknown,
+
+  // Who is making the request:
+  apiKey: asString,
+  deviceId: asString,
+  deviceToken: asOptional(asString),
+
+  // For logins:
+  loginId: asOptional(asBase64)
+})
+
+export const wasPushRequestBody = uncleaner(asJSON(asPushRequestBody))
+
+/**
+ * PUSH /v2/device/update payload.
+ */
+export const asDeviceUpdatePayload = asObject({
+  loginIds: asArray(asBase64),
+  createEvents: asOptional(asArray(asNewPushEvent), []),
+  removeEvents: asOptional(asArray(asString), [])
+})
+
+/**
+ * PUSH /v2/login/update payload.
+ */
+
+export const asLoginUpdatePayload = asObject({
+  createEvents: asOptional(asArray(asNewPushEvent), []),
+  removeEvents: asOptional(asArray(asString), [])
+})
+export type LoginUpdatePayload = $Call<typeof asLoginUpdatePayload>
+
+export const wasLoginUpdatePayload = uncleaner(asLoginUpdatePayload)
+
+// ---------------------------------------------------------------------------
+// Response types
+// ---------------------------------------------------------------------------
+
+/**
+ * A push event returned from a query.
+ */
+export const asPushEventStatus = asObject({
+  eventId: asString,
+
+  broadcastTxs: asOptional(asArray(asBroadcastTx)),
+  pushMessage: asOptional(asPushMessage),
+  recurring: asBoolean,
+  trigger: asPushTrigger,
+
+  // Status:
+  broadcastTxErrors: asOptional(asArray(asEither(asString, asNull))),
+  pushMessageError: asOptional(asString),
```

Comment:
I think this `pushMessageError` field is now two numbers, indicating the count of devices that we reached, and the count of devices that we could not reach. It should still pass your cleaner, but we may need to update this if you ever want to use the field.

Reply:
This type looks to be a string still: https://github.com/EdgeApp/edge-push-server/blob/94a54e68c1af042bb3a3f88fceda0bde343568a5/src/types/pushApiTypes.ts#L108 Does the push-server PR need updating?

Reply:
Looks like this "mystery type change" is somewhere in a working branch for the push-server. We shall revisit this tomorrow.